pkg/cluster/deploystorage.go

```diff
@@ -258,41 +258,83 @@ func (m *manager) denyAssignments(clusterSPObjectID string) *arm.Resource {
+				DenyAssignmentProperties: &mgmtauthorization.DenyAssignmentProperties{
+					DenyAssignmentName: to.StringPtr("[guid(resourceGroup().id, 'ARO cluster resource group deny assignment')]"),
+					Permissions: &[]mgmtauthorization.DenyAssignmentPermission{
+						{
+							Actions: &[]string{
+								"*/action",
+								"*/delete",
+								"*/write",
+							},
+							NotActions: &notActions,
 						},
-						NotActions: &notActions,
 					},
-				},
-				Scope: &m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID,
-				Principals: &[]mgmtauthorization.Principal{
-					{
-						ID:   to.StringPtr("00000000-0000-0000-0000-000000000000"),
-						Type: to.StringPtr("SystemDefined"),
+					Scope: &m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID,
+					Principals: &[]mgmtauthorization.Principal{
+						{
+							ID:   to.StringPtr("00000000-0000-0000-0000-000000000000"),
+							Type: to.StringPtr("SystemDefined"),
+						},
 					},
-				},
-				ExcludePrincipals: &[]mgmtauthorization.Principal{
-					{
-						ID:   &clusterSPObjectID,
-						Type: to.StringPtr("ServicePrincipal"),
+					ExcludePrincipals: &[]mgmtauthorization.Principal{
+						{
+							ID:   &clusterSPObjectID,
+							Type: to.StringPtr("ServicePrincipal"),
+						},
 					},
+					IsSystemProtected: to.BoolPtr(true),
 				},
-				IsSystemProtected: to.BoolPtr(true),
 			},
+			APIVersion: azureclient.APIVersion("Microsoft.Authorization/denyAssignments"),
 		},
-		APIVersion: azureclient.APIVersion("Microsoft.Authorization/denyAssignments"),
 	}
+
+	//if for testing denyAssignment, this if should be removed after the test is successful
+	if feature.IsRegisteredForFeature(props, "Microsoft.RedHatOpenShift/RedHatEngineering") {
+		result = append(result,
+			&arm.Resource{
+				Resource: &mgmtauthorization.DenyAssignment{
+					Name: to.StringPtr("[guid(resourceGroup().id, 'ARO cluster service principal deny assignment')]"),
+					Type: to.StringPtr("Microsoft.Authorization/denyAssignments"),
+					DenyAssignmentProperties: &mgmtauthorization.DenyAssignmentProperties{
+						DenyAssignmentName: to.StringPtr("[guid(resourceGroup().id, 'ARO cluster service principal deny assignment')]"),
+						Permissions: &[]mgmtauthorization.DenyAssignmentPermission{
+							{
+								Actions: &[]string{
+									"Microsoft.Compute/virtualMachines/deallocate/action",
+									"Microsoft.Compute/virtualMachines/powerOff/action",
+									"Microsoft.Compute/virtualMachines/simulateEviction/action",
+									"Microsoft.Compute/virtualMachineScaleSets/powerOff/action",
+									"Microsoft.Compute/virtualMachineScaleSets/deallocate/action",
+									"Microsoft.Compute/virtualMachineScaleSets/virtualMachines/powerOff/action",
+									"Microsoft.Compute/virtualMachineScaleSets/virtualMachines/deallocate/action",
+									"Microsoft.Compute/virtualMachineScaleSets/virtualMachines/simulateEviction/action",
```

Comment:
@Makdaam let's discuss this in the scrum today as post scrum. 
My question is what APIS machine-API is calling when provisioning/scaling/deleting/upgrading machines. Any of these might be on the list?
An alternative could be that we try to collect all lists of actions and allow them only, denying everything else. This is more work and more diligence but more secure from any other future "smart users".

Reply:
Since there's no official requirement that all interactions with the cloud should happen through the same api-provider what I'm afraid of is that some cloud operator starts accessing a new api or using a new action that we block - this wouldn't come up until the specific use case was used in our E2E or by a customer. For now as long as it works in OCP it will work in ARO.

Is the SP only used by machine-api-operator? I thought it also can be used by other cluster operators.

For machine API interactions the complete list is here, so I can prepare an inverted list as long as it's only the machine-api-operator with the azure provider:
https://github.com/openshift/cluster-api-provider-azure/tree/master/pkg/cloud/azure/services

Reply:
> Is the SP only used by machine-api-operator? I thought it also can be used by other cluster operators.

@Makdaam  cluster SP is also used by ingress to manage LB and in image registry. You can use
`oc get credentialsrequest -A` to see where credentials are being replicated to and according this you can assume that they are used in that namespace.

Reply:
Yes, there is Should be Loadbalancers, the ability to join into loadbalancers, some of the blobs operators for the registry. This needs updating. I would maybe even go to the lengths of checking out the code of this component and checking what kind of azure clients it uses.

And it would be good to cover these in the e2e tests too.